packages/blobstorage/indexer/indexer.go

```diff
@@ -0,0 +1,357 @@
+package indexer
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+	"github.com/ethereum/go-ethereum/ethclient"
+	blobstorage "github.com/taikoxyz/taiko-mono/packages/blobstorage"
+	"github.com/taikoxyz/taiko-mono/packages/blobstorage/bindings/taikol1"
+	"github.com/taikoxyz/taiko-mono/packages/blobstorage/pkg/repo"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/exp/slog"
+	"golang.org/x/sync/errgroup"
+)
+
+type Response struct {
+	Data []struct {
+		Index            string `json:"index"`
+		Blob             string `json:"blob"`
+		KzgCommitment    string `json:"kzg_commitment"`
+		KzgCommitmentHex []byte `json:"-"`
+	} `json:"data"`
+}
+
+// Indexer struct holds the configuration and state for the Ethereum chain listener.
+type Indexer struct {
+	beaconURL                string
+	ethClient                *ethclient.Client
+	startHeight              *uint64
+	taikoL1                  *taikol1.TaikoL1
+	db                       DB
+	blobHashRepo             blobstorage.BlobHashRepository
+	cfg                      *Config
+	wg                       *sync.WaitGroup
+	ctx                      context.Context
+	latestIndexedBlockNumber uint64
+}
+
+func (i *Indexer) InitFromCli(ctx context.Context, c *cli.Context) error {
+	cfg, err := NewConfigFromCliContext(c)
+	if err != nil {
+		return err
+	}
+
+	return InitFromConfig(ctx, i, cfg)
+}
+
+// InitFromConfig inits a new Indexer from a provided Config struct
+func InitFromConfig(ctx context.Context, i *Indexer, cfg *Config) (err error) {
+	db, err := cfg.OpenDBFunc()
+	if err != nil {
+		return err
+	}
+
+	blobHashRepo, err := repo.NewBlobHashRepository(db)
+	if err != nil {
+		return err
+	}
+
+	client, err := ethclient.Dial(cfg.RPCURL)
+	if err != nil {
+		return err
+	}
+
+	taikoL1, err := taikol1.NewTaikoL1(cfg.ContractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	i.blobHashRepo = blobHashRepo
+	i.ethClient = client
+	i.beaconURL = cfg.BeaconURL
+	i.taikoL1 = taikoL1
+	i.startHeight = cfg.StartingBlockID
+	i.db = db
+	i.wg = &sync.WaitGroup{}
+	i.ctx = ctx
+	i.cfg = cfg
+
+	return nil
+}
+
+func (i *Indexer) Start() error {
+	i.wg.Add(1)
+
+	go func() {
+		if err := i.eventLoop(i.ctx, i.latestIndexedBlockNumber); err != nil {
+			slog.Error("error in event loop", "error", err)
+		}
+	}()
+
+	return nil
+}
+
+// eventLoop runs on an interval ticker, every N seconds we will check
+// the latest processed block, and the latest header, and filter every block in between
+// for BlockProposed events, if we are not already filtering. This lets us avoid
+// unreliable subscription issues.
+func (i *Indexer) eventLoop(ctx context.Context, startBlockID uint64) error {
```

Comment:
startBlockID is not used.

Reply:
ah I should add the linter to this repo, I didn't notice it wasn't added before. it will catch things like this. I will add the linter tonight

Reply:
Ok, I already gave a [commit](https://github.com/taikoxyz/taiko-mono/pull/16353/commits/6c0d98fcedc0b25b9a7f5d0386aa0968d93794aa) and removed this one. Later you can add it again.

Reply:
Also, I don't think `filtering logic` is necessary, because `filter function` is not running Individually.
![image](https://github.com/taikoxyz/taiko-mono/assets/32827930/58d1ad7f-ad20-4957-b2ec-72d06ea8c269)

Reply:
but now the code will run filter every time the ticker ticks. what if the previous filter is still running?